core/http/server-spring/src/main/java/org/eclipse/rdf4j/http/server/repository/transaction/Transaction.java

```diff
@@ -134,13 +137,12 @@ void begin(IsolationLevel level)
 	void rollback()
 		throws InterruptedException, ExecutionException
 	{
-		Future<Boolean> result;
-		synchronized (futures) {
-			result = executor.submit(() -> {
-				txnConnection.rollback();
-				return true;
-			});
+		Future<Boolean> result = executor.submit(() -> {
+			txnConnection.rollback();
+			return true;
+		});
 
+		synchronized (futures) {
```

Comment:
Just to be clear the problem is not that `hasActiveOperations` can be called concurrently (that is by design), but that it inspects the list of Futures. It needs to lock access to that list to avoid concurrent threads modifying it _while it is looping over it_.